lib/rucio/core/request.py

```diff
@@ -527,7 +524,11 @@ def list_and_mark_transfer_requests_and_source_replicas(
         models.ReplicationRule,
         models.Request.rule_id == models.ReplicationRule.id
     ).where(
-        coalesce(models.ReplicationRule.expires_at, now) >= now
+        or_(models.ReplicationRule.child_rule_id != null(),
+            and_(models.ReplicationRule.child_rule_id == null(),
+                 models.ReplicationRule.expires_at == null()),
+            and_(models.ReplicationRule.child_rule_id == null(),
+                 models.ReplicationRule.expires_at > datetime.datetime.utcnow()))
```

Comment:
Could be simplified to

```suggestion
            and_(
                or_(models.ReplicationRule.expires_at == null(),
                    models.ReplicationRule.expires_at > datetime.datetime.utcnow()),
                models.ReplicationRule.child_rule_id == null()))
```

To avoid the duplicate `models.ReplicationRule.child_rule_id == null()`

Reply:
It all depends on whether we want to put more emphasis on code readability or query simplicity. The use of functions in SQLAlchemy for logical operators effectively results in prefix notation, which I would assume most are no familiar with. As such, I believe more than three levels of nesting should be avoided.

Otherwise, the simplest form is with `COALESCE()`, which I like, but I had received a comment in the past that it’s not readable.

```python
.where(
    or_(models.ReplicationRule.expires_at != null(),
        and_(models.ReplicationRule.child_rule_id == null(),
             coalesce(models.ReplicationRule.expires_at, now) >= now))
)
```